src/app/EventManagement.cpp

```diff
@@ -0,0 +1,867 @@
+    // to be evicted from subsequent buffers, drop it now.
+    buffer = mpEventBuffer;
+    do
+    {
+        VerifyOrExit(buffer->GetQueueSize() >= writer.GetLengthWritten(), err = CHIP_ERROR_BUFFER_TOO_SMALL);
+        if (buffer->IsFinalDestinationForPriority(opts.mpEventSchema->mPriority))
+        {
+            break;
+        }
+        else
+        {
+            buffer = buffer->GetNextCircularEventBuffer();
+        }
+    } while (true);
+
+    mBytesWritten += writer.GetLengthWritten();
+
+exit:
+    ChipLogFunctError(err);
+    if (err != CHIP_NO_ERROR)
+    {
+        *mpEventBuffer = checkpoint;
+    }
+    else if (opts.mpEventSchema->mPriority >= CHIP_CONFIG_EVENT_GLOBAL_PRIORITY)
+    {
+        CircularEventBuffer * currentBuffer = GetPriorityBuffer(opts.mpEventSchema->mPriority);
+        aEventNumber                        = currentBuffer->VendEventNumber();
+        currentBuffer->UpdateFirstLastEventTime(opts.mTimestamp);
+
+#if CHIP_CONFIG_EVENT_LOGGING_VERBOSE_DEBUG_LOGS
+        ChipLogDetail(EventLogging,
+                      "LogEvent event number: %u schema priority: %u cluster id: 0x%x event id: 0x%x sys timestamp: 0x%" PRIx64,
+                      aEventNumber, opts.mpEventSchema->mPriority, opts.mpEventSchema->mClusterId, opts.mpEventSchema->mEventId,
+                      opts.mTimestamp.mValue);
+#endif // CHIP_CONFIG_EVENT_LOGGING_VERBOSE_DEBUG_LOGS
+
+        ScheduleFlushIfNeeded(opts.mUrgent);
+    }
+
+    return err;
+}
+
+CHIP_ERROR EventManagement::CopyEvent(const TLVReader & aReader, TLVWriter & aWriter, EventLoadOutContext * apContext)
+{
+    TLVReader reader;
+    TLVType containerType;
+    CopyAndAdjustDeltaTimeContext context(&aWriter, apContext);
+    const bool recurse = false;
+    CHIP_ERROR err     = CHIP_NO_ERROR;
+
+    reader.Init(aReader);
+    err = reader.EnterContainer(containerType);
+    SuccessOrExit(err);
+
+    err = aWriter.StartContainer(AnonymousTag, kTLVType_Structure, containerType);
+    SuccessOrExit(err);
+
+    err = TLV::Utilities::Iterate(reader, CopyAndAdjustDeltaTime, &context, recurse);
+    VerifyOrExit(err == CHIP_NO_ERROR || err == CHIP_END_OF_TLV, );
+
+    err = aWriter.EndContainer(containerType);
+    SuccessOrExit(err);
+
+    err = aWriter.Finalize();
+    SuccessOrExit(err);
+
+exit:
+    ChipLogFunctError(err);
+    return err;
+}
+
+CHIP_ERROR EventManagement::EventIterator(const TLVReader & aReader, size_t aDepth, EventLoadOutContext * apEventLoadOutContext)
+{
+    CHIP_ERROR err     = CHIP_NO_ERROR;
+    const bool recurse = false;
+    TLVReader innerReader;
+    TLVType tlvType;
+    EventEnvelopeContext event;
+
+    innerReader.Init(aReader);
+    err = innerReader.EnterContainer(tlvType);
+    SuccessOrExit(err);
+
+    err = TLV::Utilities::Iterate(innerReader, FetchEventParameters, &event, recurse);
+    VerifyOrExit(event.mNumFieldsToRead == 0, err = CHIP_NO_ERROR);
+
+    err = CHIP_NO_ERROR;
```

Comment:
Please add a comment explaining why the err in ::Iterate is ignored.
It may be better to then use `(void) TLV::Utilities::Iterate...` than assigning the error and clearing it again.